subsys/net/conn_mgr/conn_mgr.c

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(conn_mgr, CONFIG_NET_CONNECTION_MANAGER_LOG_LEVEL);
+
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <errno.h>
+#include <zephyr/net/net_core.h>
+#include <zephyr/net/net_if.h>
+#include <zephyr/net/net_mgmt.h>
+
+#include <zephyr/net/conn_mgr_connectivity.h>
+#include "conn_mgr_private.h"
+
+#if defined(CONFIG_NET_TC_THREAD_COOPERATIVE)
+#define THREAD_PRIORITY K_PRIO_COOP(CONFIG_NUM_COOP_PRIORITIES - 1)
+#else
+#define THREAD_PRIORITY K_PRIO_PREEMPT(7)
+#endif
+
+/* Tracks per-iface state and event flags */
+uint16_t iface_states[CONN_MGR_IFACE_MAX];
+
+/* Tracks the total number of L4-ready ifaces */
+static uint16_t ready_count;
+
+/* Tracks the last ifaces to change state in each respective direction */
+static int last_iface_down;
+static int last_iface_up;
+
+/* Used to signal when modifications have been made that need to be responded to */
+K_SEM_DEFINE(conn_mgr_event_signal, 1, 1);
+
+/* Used to protect conn_mgr state */
+K_MUTEX_DEFINE(conn_mgr_lock);
+
+/**
+ * @brief Retrieves pointer to an iface by the index that corresponds to it in iface_states
+ *
+ * @param index - The index in iface_states to find the corresponding iface for.
+ * @return net_if* - The corresponding iface.
+ */
+static struct net_if *conn_mgr_get_if_by_index(int index)
+{
+	return net_if_get_by_index(index + 1);
```

Comment:
Why +1 here? It looks like you have now separate "index" space between this API and the net_if API.

Reply:
Some background for the index numbering. Originally network interface numbering started from 0, but then it was decided that index 0 would indicate all the network interfaces so similar was as in various Linux APIs that deal with network interfaces numbering. This means that net_if internally does +1 and -1 when figuring out the actual interface in the interface array.
So even if the index here is only used internally, I would suggest that we follow the index numbering style done in net_if. It is just quite confusing otherwise for the future developer who is looking the code and trying to understand why the index numbering differs. Also this could avoid possible hard to find bugs.

Reply:
I have now done as you suggest, Robert! @jukkar let me know if this change is satisfactory